syntax = "proto3";

service Acl {
    rpc IsPermissionedForRead (PermissionRequest) returns (PermissionResponse) {}
    rpc IsPermissionedForWrite (PermissionRequest) returns (PermissionResponse) {}
    rpc ModifyPermission (ModifyPermissionRequest) returns (ModifyPermissionResponse) {}
    rpc AddRecord (AddRecordRequest) returns (AddRecordResponse) {}
    rpc GetAllRecordsForUser (GetRecordsRequest) returns (ListOfRecords) {}
    rpc GetAllUsersForRecord (GetUsersRequest) returns (ListOfUsers) {}
}

message UserId {
    string id = 1;
}

message RecordId {
    string id = 1;
}

message PermissionRequest {
    UserId user = 1;
    RecordId record = 2;
}

message PermissionResponse {
    bool result = 1;
}

message ModifyPermissionRequest {
    UserId grantor = 1;
    UserId grantee = 2;
    RecordId record = 3;
    enum Modification {
        GRANT = 0;
        REVOKE = 1;
    }
    Modification mod = 4;
    enum Permission {
        READ = 0;
        WRITE = 1;
    }
    Permission permission = 5;
}

message ModifyPermissionResponse {
    bool result = 1;
}

message AddRecordRequest {
    UserId creator = 1;
    RecordId record = 2;
}

message AddRecordResponse {
    bool result = 1;
}

message GetRecordsRequest {
    UserId requestor = 1;
    enum Permission {
        READ = 0;
        WRITE = 1;
    }
    Permission permission = 2;
}

message ListOfRecords {
    repeated RecordId records = 1;
}

message GetUsersRequest {
    RecordId record = 1;
    enum Permission {
        READ = 0;
        WRITE = 1;
    }
    Permission permission = 2;
}

message ListOfUsers {
    repeated UserId users = 1;
}
